documentation/Semantics.md

```diff
@@ -1,206 +1,139 @@
+They should be accessed as `Symbol *` or `Symbol &` outside of the symbol
+table classes as they can't be created, copied, or moved.
+The `Symbol` class has functions and data common across all symbols, and a
+`details` field that contains more information specific to that type of symbol.
+Many symbols also have types, represented by `DeclTypeSpec`.
+Types are also owned by scopes.
+
+Name resolution happens on the parse tree in this order:
+1. Process the specification of a program unit:
+  1. Create a new scope for the unit
+  2. Create a symbol for each contained subprogram containing just the name
+  3. Process the opening statement of the unit (`ModuleStmt`, `FunctionStmt`, etc.)
+  4. Process the specification part of the unit
+2. Apply the same process recursively to nested subprograms
+3. Process the execution part of the program unit
+4. Process the execution parts of nested subprograms recursively
+
+After the completion of this phase, every `Name` corresponds to a `Symbol`
+unless an error occurred.
+
+### Rewrite parse tree
+
+The parser cannot build a completely correct parse tree without symbol information.
+This phases correct mis-parses based on symbols:
+- Array element assignments may be parsed as statement functions: `a(i) = ...`
+- Namelist group names without `NML=` may be parsed as format expressions
+- A file unit number expression may be parsed as a character variable
+
+This phase also produces an internal error if it finds a `Name` that does not
+have its `symbol` data member filled in. This error is suppressed if other
+errors have occurred because in that case a `Name` corresponding to an erroneous
+symbol may not be resolved.
+
+### Expression analysis
+
+Expressions that occur in the specification part are analyzed during name
+resolution, for example, initial values, array bounds, type parameters.
+Any remaining expressions are analyzed in this phase.
+
+For each `Variable` and top-level `Expr` (i.e. one that is not nested below
+another `Expr` in the parse tree) the analyzed form of the expression is saved
+in the `typedExpr` data member. After this phase has completed, the analyzed
+expression can be accessed using `GetExpr()` in `lib/semantics/tools.cc`.
+
+This phase also corrects mis-parses based on the result of expression analysis:
+- An expression like `a(b)` is parsed as a function reference but may need
+  to be rewritten to an array element reference (if `a` is an object entity)
+  or to a structure constructor (if `a` is a derive type)
+- An expression like `a(b:c)` is parsed as an array section but may need to be
+  rewritten as a substring if `a` is an object with type CHARACTER
+
+### Statement semantics
+
+Multiple independent checkers driven by the `SemanticsVisitor` framework
+perform the remaining semantic checks.
+By this phase, and names and expressions that can be successfully resolved
+have been. But there may be names without symbols or expressions without
```

Comment:
"have been" ?

Reply:
I meant "By this phase, **all** names and expressions that can be successfully resolved have been."